File: android/app/src/main/java/com/pokermanagement/service/SpeechInputService.kt
package com.pokermanagement.service

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.speech.RecognitionListener
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

class SpeechInputService(private val context: Context) {

    private val _transcribedText = MutableStateFlow("")
    val transcribedText: StateFlow<String> = _transcribedText

    private val _parsedPotSize = MutableStateFlow<Double?>(null)
    val parsedPotSize: StateFlow<Double?> = _parsedPotSize

    private val _parsedBetSize = MutableStateFlow<Double?>(null)
    val parsedBetSize: StateFlow<Double?> = _parsedBetSize

    private val _isListening = MutableStateFlow(false)
    val isListening: StateFlow<Boolean> = _isListening

    private var recognizer: SpeechRecognizer? = null

    fun startListening() {
        if (_isListening.value) return

        recognizer = SpeechRecognizer.createSpeechRecognizer(context).apply {
            setRecognitionListener(object : RecognitionListener {
                override fun onReadyForSpeech(params: Bundle?) {
                    _isListening.value = true
                }
                override fun onResults(results: Bundle?) {
                    val matches = results?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
                    val text = matches?.firstOrNull() ?: ""
                    _transcribedText.value = text
                    parseAmounts(text)
                    _isListening.value = false
                }
                override fun onError(error: Int) { _isListening.value = false }
                override fun onEndOfSpeech() {}
                override fun onBeginningOfSpeech() {}
                override fun onRmsChanged(rmsdB: Float) {}
                override fun onBufferReceived(buffer: ByteArray?) {}
                override fun onPartialResults(partialResults: Bundle?) {}
                override fun onEvent(eventType: Int, params: Bundle?) {}
            })
        }

        val intent = Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
            putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_FREE_FORM)
            putExtra(RecognizerIntent.EXTRA_SPEECH_INPUT_MINIMUM_LENGTH_MILLIS, 500L)
            putExtra(RecognizerIntent.EXTRA_SPEECH_INPUT_COMPLETE_SILENCE_LENGTH_MILLIS, 1500L)
            putExtra(RecognizerIntent.EXTRA_SPEECH_INPUT_POSSIBLY_COMPLETE_SILENCE_LENGTH_MILLIS, 1000L)
            putExtra(RecognizerIntent.EXTRA_MAX_RESULTS, 3)
        }

        recognizer?.startListening(intent)
    }

    fun stopListening() {
        recognizer?.stopListening()
        recognizer?.destroy()
        recognizer = null
        _isListening.value = false
    }

    private fun parseAmounts(text: String) {
        val lower = text.lowercase()

        // Pot size keywords
        val potKeywords = listOf("pot is", "pot's", "the pot")
        for (keyword in potKeywords) {
            if (lower.contains(keyword)) {
                val after = lower.substringAfter(keyword).trim()
                val amount = wordToNumber(after)
                if (amount != null) {
                    _parsedPotSize.value = amount
                    return
                }
            }
        }

        // Bet / facing keywords
        val betKeywords = listOf("bet is", "facing", "bet of", "raise to", "raise is")
        for (keyword in betKeywords) {
            if (lower.contains(keyword)) {
                val after = lower.substringAfter(keyword).trim()
                val amount = wordToNumber(after)
                if (amount != null) {
                    _parsedBetSize.value = amount
                    return
                }
            }
        }

        // Fallback: try to parse any numeric words in the utterance
        val amount = wordToNumber(lower)
        if (amount != null) {
            _parsedPotSize.value = amount
        }
    }

    private fun wordToNumber(text: String): Double? {
        // First try direct numeric parse
        val trimmed = text.trim().split(" ").firstOrNull() ?: return null
        trimmed.toDoubleOrNull()?.let { return it }

        val ones = mapOf(
            "zero" to 0, "one" to 1, "two" to 2, "three" to 3, "four" to 4,
            "five" to 5, "six" to 6, "seven" to 7, "eight" to 8, "nine" to 9,
            "ten" to 10, "eleven" to 11, "twelve" to 12, "thirteen" to 13,
            "fourteen" to 14, "fifteen" to 15, "sixteen" to 16, "seventeen" to 17,
            "eighteen" to 18, "nineteen" to 19
        )
        val tens = mapOf(
            "twenty" to 20, "thirty" to 30, "forty" to 40, "fifty" to 50,
            "sixty" to 60, "seventy" to 70, "eighty" to 80, "ninety" to 90
        )

        val words = text.trim().lowercase().split(Regex("\\s+|-"))
        var total = 0.0
        var current = 0.0

        for (word in words) {
            ones[word]?.let { current += it; return@let }
            tens[word]?.let { current += it; return@let }
            when (word) {
                "hundred" -> current *= 100
                "thousand" -> { total += current * 1000; current = 0.0 }
                "k" -> { total += current * 1000; current = 0.0 }
            }
        }
        total += current
        return if (total > 0) total else null
    }
}

File: pokerManagement/Services/SpeechInputService.swift
import Foundation
import Speech
import AVFoundation
import RegexBuilder

/// On-device voice input for pot/bet sizes via SFSpeechRecognizer (iOS 17+).
class SpeechInputService: ObservableObject {
    @Published var isListening = false
    @Published var transcribedText = ""
    @Published var parsedPotSize: Double?
    @Published var parsedBetSize: Double?
    @Published var authorizationStatus: SFSpeechRecognizerAuthorizationStatus = .notDetermined

    private let recognizer = SFSpeechRecognizer(locale: Locale(identifier: "en-US"))
    private let audioEngine = AVAudioEngine()
    private var request: SFSpeechAudioBufferRecognitionRequest?
    private var task: SFSpeechRecognitionTask?

    init() {
        SFSpeechRecognizer.requestAuthorization { [weak self] status in
            DispatchQueue.main.async { self?.authorizationStatus = status }
        }
    }

    func startListening() {
        guard authorizationStatus == .authorized, !isListening else { return }
        parsedPotSize = nil; parsedBetSize = nil; transcribedText = ""

        let req = SFSpeechAudioBufferRecognitionRequest()
        req.shouldReportPartialResults = true
        req.requiresOnDeviceRecognition = true
        self.request = req

        let node = audioEngine.inputNode
        node.installTap(onBus: 0, bufferSize: 1024, format: node.outputFormat(forBus: 0)) { buf, _ in
            req.append(buf)
        }

        do {
            let session = AVAudioSession.sharedInstance()
            try session.setCategory(.record, mode: .measurement, options: .duckOthers)
            try session.setActive(true, options: .notifyOthersOnDeactivation)
            audioEngine.prepare()
            try audioEngine.start()
        } catch { print("[Speech] audio error: \(error)"); return }

        task = recognizer?.recognitionTask(with: req) { [weak self] result, error in
            guard let self else { return }
            if let result {
                let text = result.bestTranscription.formattedString
                DispatchQueue.main.async {
                    self.transcribedText = text
                    self.parseAmounts(from: text)
                }
                if result.isFinal { self.stopListening() }
            }
            if error != nil { self.stopListening() }
        }

        DispatchQueue.main.async { self.isListening = true }
        DispatchQueue.main.asyncAfter(deadline: .now() + 10) { [weak self] in self?.stopListening() }
    }

    func stopListening() {
        guard isListening else { return }
        audioEngine.stop()
        audioEngine.inputNode.removeTap(onBus: 0)
        request?.endAudio(); task?.cancel()
        request = nil; task = nil
        DispatchQueue.main.async { self.isListening = false }
    }

    // MARK: - Parsing

    private static let words: [String: Double] = [
        "zero": 0, "one": 1, "two": 2, "three": 3, "four": 4, "five": 5,
        "six": 6, "seven": 7, "eight": 8, "nine": 9, "ten": 10,
        "eleven": 11, "twelve": 12, "thirteen": 13, "fourteen": 14, "fifteen": 15,
        "sixteen": 16, "seventeen": 17, "eighteen": 18, "nineteen": 19,
        "twenty": 20, "thirty": 30, "forty": 40, "fifty": 50,
        "sixty": 60, "seventy": 70, "eighty": 80, "ninety": 90,
        "hundred": 100, "thousand": 1000
    ]

    private func parseAmounts(from text: String) {
        let low = text.lowercased()
        if let v = extract(after: ["the pot is", "pot is", "pot's", "pot"], in: low) { parsedPotSize = v }
        if let v = extract(after: ["facing a bet of", "bet is", "bet of", "facing", "bet"], in: low) { parsedBetSize = v }
    }

    private func extract(after prefixes: [String], in text: String) -> Double? {
        for p in prefixes {
            guard let r = text.range(of: p) else { continue }
            let after = String(text[r.upperBound...]).trimmingCharacters(in: .whitespaces)
            let digits = Regex { Capture { OneOrMore(.digit) } }
            if let m = after.prefixMatch(of: digits), let v = Double(m.1) { return v }
            if let v = parseWords(after) { return v }
        }
        return nil
    }

    private func parseWords(_ text: String) -> Double? {
        var total = 0.0, cur = 0.0, found = false
        for w in text.split(separator: " ").map({ String($0).lowercased() }) {
            guard let v = Self.words[w] else { if found { break } else { continue } }
            found = true
            if v == 100 { cur = (cur == 0 ? 1 : cur) * 100 }
            else if v == 1000 { cur = (cur == 0 ? 1 : cur) * 1000; total += cur; cur = 0 }
            else { cur += v }
        }
        total += cur
        return found ? total : nil
    }
}
